plinth/modules/ejabberd/views.py

```diff
@@ -33,8 +37,52 @@ class EjabberdServiceView(ServiceView):
     template_name = 'ejabberd.html'
     description = ejabberd.description
     diagnostics_module_name = 'ejabberd'
+    form_class = EjabberdForm
+
+    def get_initial(self):
+        initdict = super().get_initial()
+        initdict.update({'MAM_enabled': self.is_MAM_enabled()})
+        return initdict
 
     def get_context_data(self, *args, **kwargs):
+        """Add service to the context data."""
         context = super().get_context_data(*args, **kwargs)
-        context['domainname'] = ejabberd.get_domainname()
         return context
+
+    def form_valid(self, form):
+        """Enable/disable a service and set messages."""
+        old_status = form.initial
+        new_status = form.cleaned_data
+        app_same = old_status['is_enabled'] == new_status['is_enabled']
+        mam_same = old_status['MAM_enabled'] == new_status['MAM_enabled']
+
+        if app_same and mam_same:
+            # TODO: find a more reliable/official way to check whether the
+            # request has messages attached.
+            if not self.request._messages._queued_messages:
+                messages.info(self.request, _('Setting unchanged'))
+        elif not app_same:
+            if new_status['is_enabled']:
+                self.service.enable()
+                messages.success(self.request, _('Application enabled'))
+            else:
+                self.service.disable()
+                messages.success(self.request, _('Application disabled'))
+
+        if not mam_same:
+            # note ejabberd action "enable" or "disable" restarts, if running
+            if new_status['MAM_enabled']:
+                actions.superuser_run('ejabberd', ['mam', 'enable'])
+                messages.success(self.request,
+                                 _('Message Archive Management enabled'))
+            else:
+                actions.superuser_run('ejabberd', ['mam', 'disable'])
+                messages.success(self.request,
+                                 _('Message Archive Management disabled'))
+
+        return super(ServiceView, self).form_valid(form)
```

Comment:
Hi, thanks for checking these things.
It's true that ServiceView.form_valid() is quite similar, however it's necessary to overwrite the parent function here, since the check of MAM's status is needed to e.g. know that 'Setting unchanged' is true.
That's also why at the end, even the *grandparent* function gets called instead of just `super().form_valid(form)`, because this would add some wrong checks (and thus messages).
So I guess another option would be not to inherit from ServiceView, but I think that would mean even more code duplication.